src/assertions.jl

```diff
@@ -0,0 +1,41 @@
+"""
+    Assertion
+
+An Assertion stores settings for checking the validity of a `AbstractArray` or `Tables.table` containing missing values.
+New assertions are expected to subtype `Impute.Assertion` and, at minimum,
+implement the `assert(data::AbstractVector{Union{T, Missing}}, ::<MyAssertion>)` method.
+"""
+abstract type Assertion end
+
+
+"""
+    assert(data, a::Assertion; kwargs...)
+
+If the assertion `a` fails then an `AssertionError` is thrown, otherwise the `data`
+provided is returned without mutation.
+
+# Keywords
+* `dims`: The dimension to apply the assertion along (e.g., observations dim)
+"""
+assert
+
+# A couple fallbacks for matrices and tables
+function assert(data::AbstractMatrix, a::Assertion; dims=1)
+    for var in varwise(data; dims=dims)
+        assert(var, a)
```

Comment:
Do we want to pass on any leftover kwargs?
Or is that just gold-plating?
```suggestion
function assert(data::AbstractMatrix, a::Assertion; dims=1; kwargs...)
    for var in varwise(data; dims=dims)
        assert(var, a; kwargs...)
```

Reply:
Hmm, we don't currently have a use-case for that, but I don't see any harm. I'm all for `gold-plating` :)